### PR TITLE
Desnet network demo ready to be reviewed

### PR DESCRIPTION
**Pull request by jdenholm**
Would mind playing around and reviewing this simple classification demo in [``demos/dense_network_classification.ipynb``](https://github.com/jdenholm/TorchTools/blob/dn-demo/demos/dense_network_classification.ipynb) on branch ``dn-demo``, please?

Things to do:

- [x] Make it clear that the demo environment is different from the regular one.
- [x] Add print-outs showing which physical characteristics of the tumours are being used.
- [x] Add text explaining the tensors' dimensionality print-outs.
- [x] Consider commenting the test and validation functions.
- [x] Remove random empty cell.
- [x] Expand print-outs of the dataset to introduce/explain it more clearly.
- [x] References for things like the optimiser and dataset, etc.
- [x] Add more text on the role of a test set.

### CODE REVIEW

**Comment by ah3918** [comment]
Hi Jim, thanks for asking me to review! I think it's great,  just the right length and simple enough. Minor points (and apologies in advance if these comments demonstrate my lack of ML and python knowledge):

- Sklearn, pandas and matplotlib missing from yaml file (although I suppose users interested in this should have these already). 
- Very minor but maybe add a print statement of the data_keys to show the user what kind of characteristics the breast cancer dataset has.
- Again, minor and probably showing my lack of knowledge (my python is awful!), maybe explain what “torch.Size” means (block 5 , last print statement)? Or preface the demo with “this assumes knowledge of pytorch” etc. ?
- I’m not very familiar with the computational strain of ML/DL methods (if that wasn’t obvious already!). Is it worth adding a note above the CPU/GPU selection about keeping resource need in mind when running classification? I’m sure something like this needs very little.
- Maybe add comments throughout the test/train functions? Depends on your target audience ofc.
- Empty cell before plotting cell

Hope this helps!

**Comment by jatkinson1000** [comment]
Some points (as discussed) on CI:
- Perform black, docschecks, and linting once in a separate file rather than for every setup as the results are the same.
- Add `if: always()` to force all checks to run rather than falling at the first missed hurdle.
- Consider restricting when the checks are run (merges to main, pushes to PRs etc rather than for everything).

On the jupyter notebook (in addition to @ah3918's comments above):
- It's not immediately obvious to me what type the dataset is. Knowing would help me understand what's being done/what to google for.
- Might be nice to link to external references for some things (e.g. I'm wondering what the 'Adam' optimiser is).
- You don't bother with a testing set, but if this is a tutorial it might be worth doing things 'properly'.

**Comment by jdenholm** [comment]
@jatkinson1000, I'm going to merge this PR and address the CI stuff as a separate issue, but I completely agree with you about it.
